frontend/src/components/ui/BuyMenu.jsx

```diff
@@ -148,12 +151,15 @@ export default function BuyMenu({game}) {
 
                 game.gameMapOverrides.clear();
 
+                let tileType;
                 if (selectedItem.name == TERRAFORM_FILL) {
                     z += 1;
                     gameMap.addTile(x, y, z, new Tile(x, y, z, TileType.Stone));
                 } else if (selectedItem.name == TERRAFORM_EXCAVATE) {
+                    tileType = gameMap.getTile(x, y, z).type;
                     gameMap.removeTile(x, y, z);
                 }
+                setUndoStack(prevStack => [...prevStack, {x, y, z, ...terraform, tileType}]);
```

Comment:
I would just make the keys explicit instead of having the invisible dependency, where we rely on `terraform` having the right values defined. And that way you can name them whatever you want for the undo logic.